woocommerce/woocommerce-gateway-stripe
# Fixes invalid recurring shipping method error when purchasing multiple subscriptions with Apple Pay or Google Pay

Fixes #2837

Background/relevant links:
- A HVM was impacted by this same issue on WooPayments and so I fixed the same problem in both Stripe and WooPayments
- WooPayments PR here: https://github.com/Automattic/woocommerce-payments/pull/8618
- Slack thread: p1712735606268559-slack-C7U3Y3VMY

## Changes proposed in this Pull Request:







When attempting to purchase multiple subscriptions of varying periods with Apple/Google Pay, you will get an "Invalid recurring shipping method." error:

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/2e74b3d5-34a7-4ddb-938a-51cbb66decf5)

This issue is caused by our [`WC_Stripe_Payment_Request::calculate_shipping()`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/df901e975a3a231d9ce265ff0299001af8a0be54/includes/payment-methods/class-wc-stripe-payment-request.php#L1725) function wiping the chosen shipping methods when it recalculates shipping packages and available shipping methods.

During the `process_checkout()` call, Woo Subscriptions validates the checkout data and was throws an error because the recurring shipping methods are no longer set in the `WC()->session`.

This PR fixes these issues by restoring all chosen recurring shipping methods that were wiped.

## Testing instructions





1. Install and activate the Woo Subscriptions extension
2. Enable Apple Pay and Google Pay
3. Create two subscription products, one that renews monthly and another that renews weekly
4. Add both of these products to your cart, and then visit the checkout page
5. While on `develop`, click on the Google Pay PRB.
6. When you attempt to confirm the Google Pay window, you should see the error in the screenshot above.
7. While on this branch, the payment will be processed and you will have 2 new subscriptions with the correct shipping methods.

---


**Post merge**
